src/lib.rs

```diff
@@ -0,0 +1,850 @@
+            self.repetition_counter = repetition_counter;
+            self
+        }
+
+        pub fn enable_repetition_interrupt(mut self) -> Self {
+            self.enable_repetition_interrupt = true;
+            self
+        }
+
+        pub fn eev_cfg(mut self, eev_cfg: EevCfgs<$TIMX>) -> Self {
+            self.eev_cfg = eev_cfg;
+            self
+        }
+    };
+}
+
+// Implement PWM configuration for timer
+macro_rules! hrtim_hal {
+    ($($TIMX:ident: $($out:ident)*,)+) => {
+        $(
+            impl HrPwmAdvExt for $TIMX {
+                type PreloadSource = PreloadSource;
+
+                fn pwm_advanced<PINS>(
+                    self,
+                    pins: PINS,
+                    _rcc: &mut Rcc,
+                ) -> HrPwmBuilder<Self, Pscl128, Self::PreloadSource, PINS>
+                where
+                    PINS: ToHrOut<$TIMX>,
+                {
+                    // TODO: That 32x factor... Is that included below, or should we
+                    // do that? Also that will likely risk overflowing u32 since
+                    // 170MHz * 32 = 5.44GHz > u32::MAX.Hz()
+                    //let clk = HertzU64::from(HRTIM_COMMON::get_timer_frequency(&rcc.clocks)) * 32;
+
+                    HrPwmBuilder {
+                        _tim: PhantomData,
+                        _prescaler: PhantomData,
+                        pins,
+                        timer_mode: HrTimerMode::Continuous,
+                        fault_enable_bits: 0b000000,
+                        fault1_bits: 0b00,
+                        fault2_bits: 0b00,
+                        counting_direction: HrCountingDirection::Up,
+                        //base_freq: clk,
+                        count: CountSettings::Period(u16::MAX),
+                        preload_source: None,
+                        enable_push_pull: false,
+                        interleaved_mode: InterleavedMode::Disabled,
+                        repetition_counter: 0,
+                        deadtime: None,
+                        enable_repetition_interrupt: false,
+                        eev_cfg: EevCfgs::default(),
+                        out1_polarity: Polarity::ActiveHigh,
+                        out2_polarity: Polarity::ActiveHigh,
+                    }
+                }
+            }
+
+            impl<PSCL, PINS>
+                HrPwmBuilder<$TIMX, PSCL, PreloadSource, PINS>
+            where
+                PSCL: HrtimPrescaler,
+                PINS: ToHrOut<$TIMX>,
+            {
+                pub fn finalize(self, _control: &mut HrPwmControl,
+                    #[cfg(feature = "stm32f3")]
+                    moder: &mut <PINS::GpioX as hal::gpio::marker::GpioStatic>::MODER,
+                    #[cfg(feature = "stm32f3")]
+                    otyper: &mut <PINS::GpioX as hal::gpio::marker::GpioStatic>::OTYPER,
+                    #[cfg(feature = "stm32f3")]
+                    afr: &mut PINS::Afr
```

Comment:
Turned out to be quite some differences in how the gpio types work in g4 vs f3 hals. I made an ugly fix for now just to get it to compile but there is a lot if room for improvement there.

The main difference that affects this is how pins are put into alternate mode. The f3 hal requires exclusive access to the gpio blocks shared registers (moder, otyper and afr) which makes sense since it will in fact modify them.

```rust
let mut gpioa = dp.GPIOA.split(&mut rcc.ahb);
let pa4 = gpioa
    .pa4
    .into_af_push_pull(&mut gpioa.moder, &mut gpioa.otyper, &mut gpioa.afrl);
```

The g4 hal however, does not require this and just conjures access to these registers internally(likely not sound?).

```rust
let gpioa = dp.GPIOA.split(&mut rcc);
let pin = gpioa.pa8.into_alternate();
```

It is recommended that the pins are connected the HRTIM peripheral only as the last step of initialization to prevent any glitches from occuring on the pins. In my mind, this means that the `into_af_push_pull` and `into_alternate` methods should be called by the hrtim init, not by the user. This means that we need to pass in those registers in the f3 case. However there is also the use case where the peripheral is used without any output pins...